data_iterator.py

```diff
@@ -51,22 +56,35 @@ def create_padded_batch(state, x):
             X[triple_length:, idx] = state['eos_sym']
 
         # Initialize Xmask column with ones in all positions that
-        # were just set in X
+        # were just set in X. 
+        # Note: if we need mask to depend on tokens inside X, then we need to 
+        # create a corresponding mask for X_reversed and send it further in the model
         Xmask[:triple_length, idx] = 1.
 
-        # Find start of last utterance
+        # Reverse all utterances
         eos_indices = numpy.where(X[:, idx] == state['eos_sym'])[0]
+        X_reversed[:triple_length, idx] = x[0][idx][:triple_length]
+        prev_eos_index = -1
+        for eos_index in eos_indices:
+            X_reversed[(prev_eos_index+2):eos_index, idx] = (X_reversed[(prev_eos_index+2):eos_index, idx])[::-1]
+            prev_eos_index = eos_index
+            if prev_eos_index > triple_length:
+                break
```

Comment:
@sordonia I just did a quick test, and it takes less than 50 ms to compute create_padded_batch() function for a batch of size 20 (including the standard data and mask). Mean while, training on a batch size 40 in total takes about 1-2s. This means create_padded_batch()  takes up about 5% of the training time, so we don't have to worry about its performance right now...